types/src/proof/definition.rs

```diff
@@ -303,6 +304,62 @@ impl AccumulatorConsistencyProof {
     pub fn siblings(&self) -> &[HashValue] {
         &self.siblings
     }
+
+    /// Verifies the proof. The verification consists of two steps. First, the frozen subtree roots
+    /// should match `old_root_hash` and `old_num_leaves`. Second, after combining the frozen
+    /// subtree roots with the siblings we should be able to obtain `new_root_hash`.
+    pub fn verify<H: CryptoHasher>(
+        &self,
+        old_root_hash: HashValue,
+        old_num_leaves: u64,
+        new_root_hash: HashValue,
+        new_num_leaves: u64,
+    ) -> Result<()> {
+        ensure!(
+            old_num_leaves <= new_num_leaves,
+            "Old accumulator ({} leaves) is bigger than new accumulator ({} leaves).",
+            old_num_leaves,
+            new_num_leaves,
+        );
+
+        let old_accumulator =
+            Accumulator::<H>::new(self.frozen_subtree_roots.clone(), old_num_leaves)?;
+        ensure!(
+            old_accumulator.root_hash() == old_root_hash,
+            "Frozen subtree roots in the proof do not match old root hash. \
+             Actual root hash: {:x}. Expected root hash: {:x}.",
+            old_accumulator.root_hash(),
+            old_root_hash,
+        );
+
+        if old_num_leaves == 0 {
+            ensure!(
```

Comment:
I think there should be an explicit test case for this.

Reply:
Also it wouldn't hurt to test the error paths to make sure we actually get errors.  I assume people will get broken  proofs from bad clients at some point and try to verify them.